src/services/userSettings.js

```diff
@@ -0,0 +1,193 @@
+import { Op } from 'sequelize';
+import { USER_SETTINGS } from '../constants';
+import {
+  sequelize, User, UserSettings, UserSettingOverrides,
+} from '../models';
+
+const baseSearch = (userId) => ({
+  attributes: [
+    'key',
+    [sequelize.fn('COALESCE', sequelize.col('"userSettingOverrides"."value"'), sequelize.col('"UserSettings"."default"')), 'value'],
+  ],
+  include: [
+    {
+      attributes: [],
+      model: UserSettingOverrides,
+      as: 'userSettingOverrides',
+      where: { userId: { [Op.eq]: userId } },
+      required: false,
+    },
+  ],
+});
+
+/**
+ * Returns an object of all setting keys, their default values and their id.
+ * @returns {Promise<{ [key]: { defaultValue: any, userSettingId: number, key: string }}}>}
+ */
+export const getDefaultSettings = async () => {
+  const out = await UserSettings.findAll({});
+
+  return out.map(({ dataValues: { key, default: defaultValue, id: userSettingId } }) => ({
+    key, defaultValue, userSettingId,
+  })).reduce((acc, { key, defaultValue, userSettingId }) => {
+    acc[key] = { userSettingId, defaultValue, key };
+    return acc;
+  }, {});
+};
+
+/**
+ * Given an array of { key, value } pairs, saves these settings to the database.
+ * @param {number} userId
+ * @param {{key: string, value: any}[]} pairs
+ */
+export const saveSettings = async (userId, pairs) => {
+  const defaults = await getDefaultSettings();
+
+  const save = pairs.reduce((acc, { key, value }) => {
+    if (defaults[key] && defaults[key].defaultValue !== value) {
+      return [...acc, { key, value, userSettingId: defaults[key].userSettingId }];
+    }
+    return acc;
+  }, []);
+
+  const del = pairs.reduce((acc, { key, value }) => {
+    if (defaults[key] && defaults[key].defaultValue === value) {
+      return [...acc, { key, value, userSettingId: defaults[key].userSettingId }];
+    }
+    return acc;
+  }, []);
+
+  const settingIds = del.map(({ userSettingId }) => userSettingId)
+    .concat(save.map(({ userSettingId }) => userSettingId));
+
+  return Promise.all([
+    UserSettingOverrides.destroy({ where: { userId, userSettingId: { [Op.in]: settingIds } } }),
+    UserSettingOverrides.bulkCreate(save.map(({ userSettingId, value }) => ({
+      userId,
+      userSettingId,
+      value: sequelize.cast(JSON.stringify(value), 'jsonb'),
+    }))),
+  ]);
+};
```

Comment:
I know we use this pattern in other places (implemented during an accelerated initial MVP schedule), but ideally, we would change it to updates, if possible. Additionally, with the implementation of the audit log, we end up with a less readable output.

Reply:
I refactored this to `update` where possible, `insert` where necessary and `delete` just as it did before. I think this should result in more readable audit log entries. Does it look okay? c2107d60

Reply:
Thanks. The code needs to be modified to also include "key" in the `updateable` or the filtering can use `userSettingId `instead.

Sequelize also offers `findOrCreate `or `upsert `that could be helpful here.

Reply:
thanks for catching this! I believe this is now fixed. I wrote some additional tests to verify this behavior works as intended.